activesupport/lib/active_support/core_ext/class/attribute.rb

```diff
@@ -92,50 +92,65 @@ def class_attribute(
     default: nil
   )
-    attrs.each do |name|
-      singleton_class.silence_redefinition_of_method(name)
-      define_singleton_method(name) { default }
+    class_method_def = []
+    method_def = []
 
-      singleton_class.silence_redefinition_of_method("#{name}?")
-      define_singleton_method("#{name}?") { !!public_send(name) } if instance_predicate
+    attrs.each do |name|
+      unless name.is_a?(Symbol) || name.is_a?(String)
+        raise TypeError, "#{name.inspect} is not a symbol nor a string"
+      end
 
-      ivar = "@#{name}".to_sym
+      class_method_def <<
+        "silence_redefinition_of_method(:#{name})" <<
+        "def #{name}" << # In case the method exists and is not public
+        "end"
 
-      singleton_class.silence_redefinition_of_method("#{name}=")
-      define_singleton_method("#{name}=") do |val|
-        redefine_singleton_method(name) { val }
+      class_method_def <<
+        "silence_redefinition_of_method(:#{name}=)" <<
+        "def #{name}=(value)" <<
+        "  redefine_method(:#{name}) { value } if singleton_class?" <<
+        "  redefine_singleton_method(:#{name}) { value }" <<
+        "  value" <<
+        "end"
 
-        if singleton_class?
-          class_eval do
-            redefine_method(name) do
-              if instance_variable_defined? ivar
-                instance_variable_get ivar
-              else
-                singleton_class.send name
-              end
-            end
-          end
-        end
-        val
+      if instance_predicate
+        class_method_def <<
+          "silence_redefinition_of_method(:#{name}?)" <<
+          "def #{name}?" <<
+          "  !!self.#{name}" <<
+          "end"
       end
 
       if instance_reader
-        redefine_method(name) do
-          if instance_variable_defined?(ivar)
-            instance_variable_get ivar
-          else
-            self.class.public_send name
-          end
-        end
+        method_def <<
+          "silence_redefinition_of_method(:#{name})" <<
+          "def #{name}" <<
+          "  defined?(@#{name}) ? @#{name} : self.class.#{name}" <<
+          "end"
 
-        redefine_method("#{name}?") { !!public_send(name) } if instance_predicate
+        if instance_predicate
+          method_def <<
+            "silence_redefinition_of_method(:#{name}?)" <<
+            "def #{name}?" <<
+            "  !!self.#{name}" <<
+            "end"
+        end
       end
 
       if instance_writer
-        redefine_method("#{name}=") do |val|
-          instance_variable_set ivar, val
-        end
+        method_def <<
+         "silence_redefinition_of_method(:#{name}=)" <<
+         "attr_writer :#{name}"
       end
     end
+
+    location = caller_locations(1, 1).first
+    file, line = location.path, location.lineno
+
+    class_eval(["class << self", *class_method_def, "end", *method_def].join(";"), file, line)
+
+    attrs.each do |name|
+      self.public_send("#{name}=", default)
+    end
```

Comment:
Find the code significantly tougher to read, so here's a potential, albeit only tested in parts, alternative:

```ruby
def class_attribute(*attrs, instance_accessor: true, instance_reader: instance_accessor, instance_writer: instance_accessor, instance_predicate: true, default: nil)
  if name = attrs.find { |name| name.is_a?(Symbol) || name.is_a?(String) }
    raise TypeError, "#{name.inspect} is not a symbol nor a string"
  end

  class_methods, methods = [], []
  attrs.each do |name|
    class_methods << "silence_redefinition_of_method def #{name}; end" # In case the method exists and is not public

    methods << <<~end_of_method if instance_reader
      silence_redefinition_of_method def #{name}
        defined?(@#{name}) ? @#{name} : self.class.#{name}
      end
    end_of_method


    class_methods << <<~end_of_method
      silence_redefinition_of_method def #{name}=(value)
        redefine_method(:#{name}) { value } if singleton_class?
        redefine_singleton_method(:#{name}) { value }
        value
      end
    end_of_method

    methods << <<~end_of_method if instance_writer
      silence_redefinition_of_method(:#{name}=)
      attr_writer :#{name}
    end_of_method


    if instance_predicate
      class_methods << "silence_redefinition_of_method def #{name}?; !!self.#{name}; end"
      methods       << "silence_redefinition_of_method def #{name}?; !!self.#{name}; end"
    end
  end

  location = caller_locations(1, 1).first
  class_eval(["class << self", *class_methods, "end", *methods].join.tr("\n", ";"), location.path, location.lineno)

  attrs.each { |name| public_send("#{name}=", default) }
end
```

Think it's clearer if we group by method kind instead of class then instance. We can also take advantage of `def` returning the method name as a symbol and save a line in most cases (`attr_writer` returns `nil`). Also think it's clearer if the type check is hoisted up top, despite the extra loop. Think there's a significant win with heredoc + tailing ifs.

Reply:
@kaspth I'm fine with using `.tr("\n", ";")`, but then we should probably apply the same change to `Module#delegate` for consistency's sake. 

cc @rafaelfranca WDYT?

Reply:
I'm fine with that.

Reply:
Done. With a few adjustments.

I'll see later for Module#delegate.